File: code/cpp/ch10/practice/challenge_01_fibonacci_three_ways.cpp
/*
 * Challenge 1: Fibonacci Three Ways
 * ====================================
 * Chapter 10: The Magic of Recursion
 *
 * PROBLEM:
 *   Implement the nth Fibonacci number THREE different ways:
 *     solve_naive(n) -- pure recursion (exponential time)
 *     solve_memo(n)  -- recursion with memoization (linear time)
 *     solve_iter(n)  -- iterative with two variables (linear time)
 *     solve(n)       -- calls solve_iter (the best approach)
 *
 * EXAMPLES:
 *   solve_naive(10) -> 55
 *   solve_memo(30)  -> 832040
 *   solve_iter(30)  -> 832040
 *   solve(30)       -> 832040
 *
 * CONSTRAINTS:
 *   0 <= n <= 50 (for memo and iter)
 *   0 <= n <= 30 (for naive -- too slow beyond this!)
 *   Use long long to handle large values.
 *
 * INSTRUCTIONS:
 *   Replace all four function bodies with your solutions.
 */

#include <iostream>
#include <unordered_map>
using namespace std;

long long solve_naive(int n) {
    // TODO: Pure recursion (no optimization)
    return 0;
}

long long solve_memo(int n) {
    // TODO: Recursion with memoization (unordered_map)
    return 0;
}

long long solve_iter(int n) {
    // TODO: Iterative with two variables
    return 0;
}

long long solve(int n) {
    // TODO: Call solve_iter
    return solve_iter(n);
}

// -- Do not change anything below this line --------------------------
int main() {
    int n;
    cin >> n;
    cout << solve(n) << endl;
    return 0;
}

File: code/java/ch10/practice/Challenge01FibonacciThreeWays.java
package ch10.practice;

import java.util.*;

/**
 * Challenge 01: Fibonacci — Three Ways
 * ==============================
 * Chapter 10: The Magic of Recursion
 *
 * PROBLEM: Implement Fibonacci three ways to compare approaches:
 *   1. solveNaive(n)  — pure recursion (exponential time)
 *   2. solveMemo(n)   — recursion with memoization (linear time)
 *   3. solveIter(n)   — iterative with two variables (linear time, O(1) space)
 *   4. solve(n)       — calls solveIter (the best approach)
 *
 * EXAMPLES:
 *   solveNaive(10) = 55
 *   solveMemo(30)  = 832040
 *   solveIter(30)  = 832040
 *
 * CONSTRAINTS: 0 <= n <= 45 for memo/iter, 0 <= n <= 30 for naive
 *
 * INSTRUCTIONS: Replace the bodies of all four methods with your solutions.
 */
public class Challenge01FibonacciThreeWays {
    public static long solveNaive(int n) {
        // TODO: Replace this with your solution
        return 0;
    }

    public static long solveMemo(int n) {
        // TODO: Replace this with your solution
        return 0;
    }

    public static long solveIter(int n) {
        // TODO: Replace this with your solution
        return 0;
    }

    public static long solve(int n) {
        // TODO: Replace this with your solution (use solveIter)
        return 0;
    }

    // ── Do not change anything below this line ──────────────────────
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = Integer.parseInt(sc.nextLine().trim());
        System.out.println(solve(n));
        sc.close();
    }
}
